src/app/navigation/reducer.ts

```diff
@@ -15,6 +19,15 @@ export default (location: Location): Reducer<State, AnyAction> => (state = addQu
   switch (action.type) {
     case getType(actions.locationChange):
       return addQuery(action.payload.location);
+    case getType(deleteHighlight): {
+      const scrollTarget = getScrollTargetFromQuery(state.query, state.hash);
+      if (scrollTarget && isHighlightScrollTarget(scrollTarget) && scrollTarget.id === action.payload) {
+        const window = assertWindow();
+        window.history.replaceState(null, '', window.location.origin + window.location.pathname);
+        return {...state, search: '', hash: '', query: {}};
+      }
+      return state;
+    }
```

Comment:
Ideally, reducers should be pure functions of actions and state.

Reply:
Are you concerned about using `window.history` here or do you mean that this part of code should be extracted to the separated function?

The thing is that we don't really have to clear the history state here - clearing search, hash and query is enough, however, then there is a scrollTarget in the url

I was thinking about dispatching `locationChange` in deleteHighlight hook, but this is async action so the error alert will be already displayed + we would have block somehow this locationChange so we don't load a page again

Reply:
Pure functions can't have side effects and modifying `window.history` is one.

Reply:
yea putting this replaceState in the reducer is kinda bad, really it should be in a redux hook or something

Reply:
the problems is a follows:
- we want to remove the scroll target from the URL when a user removes the highlight that it was pointing to
- we have to do this synchronously because if we not then highlightManager will know that there is a scrollTarget but it will not find the highlight and will throw the error
- to stop the error from happening we have to clear the navigation's state as in the example above
- there is still a scroll target in the URL
- we can't clear it in hook because it is running after the action updated the state so we no longer know what is the current scroll target
- we could parse `window.location.search` manually in the hook but this seems weird + it is splitting the code for clearing the scroll target into even more places

This could be solve by using `redux-thunk` approach that I mentioned in another pr https://github.com/openstax/rex-web/pull/822#issuecomment-694626803  

Any ideas?

Reply:
the solution is to use two actions, you'd have to do this with redux-thunk as well because thunk actions are not reduced.

you use one action, which i sometimes prefix with "request" like `requestDeleteHighlight` which may or may not be reduced at all, theoretically it could create some kind of loading state. that action will trigger a redux hook which does whatever side effect you need and then dispatch `receiveDeleteHighlight` indicating that the action is complete and you can reduce that action to update the state.

typically i'd do the api call in response to the `request` action and dispatch the result with the `receive` payload, but this does not fit with the current eager/optimistic design of the delete highlight interaction so in order to maintain your current behavior you might have to hook into `request` to do your history update and again into `receive` to do the api call